src/builder/reduce.js

```diff
@@ -61,23 +65,34 @@ module.exports = function (file, ipld, options) {
     const links = leaves.map((leaf) => {
       f.addBlockSize(leaf.leafSize)
 
-      return new DAGLink(leaf.name, leaf.size, leaf.multihash)
+      let cid = leaf.cid
+
+      if (!cid) {
+        // we are an intermediate node
+        cid = new CID(options.cidVersion, 'dag-pb', leaf.multihash)
+      }
+
+      return new DAGLink(leaf.name, leaf.size, cid.buffer)
     })
 
     waterfall([
       (cb) => DAGNode.create(f.marshal(), links, options.hashAlg, cb),
       (node, cb) => {
-        if (options.onlyHash) return cb(null, node)
-
-        let cid = new CID(node.multihash)
+        const cid = new CID(options.cidVersion, 'dag-pb', node.multihash)
 
-        if (options.cidVersion === 1) {
-          cid = cid.toV1()
+        if (options.onlyHash) {
+          return cb(null, {
+            node, cid
+          })
         }
 
-        ipld.put(node, { cid }, (err) => cb(err, node))
+        ipld.put(node, {
+          cid
+        }, (err) => cb(err, {
+          node, cid
+        }))
       }
-    ], (err, node) => {
+    ], (err, {node, cid}) => {
```

Comment:
You can't destructure like this in a callback.

If err, then the result will be undefined/null and will try to destructure the properties anyway and throw.